### ISSUE TITLE
Stream: Unhandled Exception on Disconnection

### ISSUE THREAD

**Issue by nikop** [linvi/tweetinvi#290]
https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Streams/StreamTask.cs#L245

Exception (System.IO.IOException) raised by "ReadLineAsync" isn't handled, which will cause process to exit.

**Comment by linvi** [comment]
As you can see `GetJsonResponseFromReader` is invoked only in `try/catch` environment.
https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Streams/StreamTask.cs#L118

The problem you have is probably another problem.Could you share you callstack.

Thanks

**Comment by nikop** [comment]
`System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.IO.IOException: Unable to read data from the transport connection: An established connection was aborted by the software in your host machine. ---> System.Net.Sockets.SocketException: An established connection was aborted by the software in your host machine
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.ConnectStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.Http.HttpClientHandler.WebExceptionWrapperStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Http.DelegatingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamReader.ReadBuffer()
   at System.IO.StreamReader.ReadLine()
   at AsyncExtensions.<>c__DisplayClassf.<ReadLineAsync>b__e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.IO.IOException: Unable to read data from the transport connection: An established connection was aborted by the software in your host machine. ---> System.Net.Sockets.SocketException: An established connection was aborted by the software in your host machine
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.ConnectStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.Http.HttpClientHandler.WebExceptionWrapperStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Http.DelegatingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamReader.ReadBuffer()
   at System.IO.StreamReader.ReadLine()
   at AsyncExtensions.<>c__DisplayClassf.<ReadLineAsync>b__e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()<---`

I believe If timeout happens requestTask is never awaited, which results expection not being catched

**Comment by linvi** [comment]
I will investigate and let you know if I can reproduce it.

**Comment by linvi** [comment]
Could you please share your code as I cannot reproduce your problem and the exception is for me handled properly by the `StreamStopped` event.

**Comment by nikop** [comment]
By disconnected I mean when it happens by error, not user stopping.

`static public async void Setup()
{
	_stream = Stream.CreateFilteredStream();

	foreach (var user in _posters)
	{
		_stream.AddFollow(User.GetUserFromScreenName(user.Key));
	}

	_stream.MatchingTweetReceived += (sender, argsx) =>
	{
		if (argsx.Tweet.IsRetweet)
			return;

		GetLogger("Twitter").Write($"Got Tweet from {argsx.Tweet.CreatedBy.Name}: {argsx.Tweet.Text}");
	};

	try
	{
		await _stream.StartStreamMatchingAnyConditionAsync();
	}
	catch (Exception ex)
	{
		LogError(ex);
	}
}`

**Comment by linvi** [comment]
I am sorry, I have copy/pasted your code and the error is properly handled.
Can you confirm that you are using version 1.0. Also can you confirm that the LoggedError comes from the stream as I cannot see that in the StackTrace.

I am bit curious about this issue as the code to catch such problem is in place.
Please use the `StreamStopped` event and check if you receive some information from there.

**Comment by nikop** [comment]
Yes, I'm using latest version and there's no other place that could cause it. I've gone over all code and ensured all Tasks are properly handled.

My belief is this is related to timing of events, ie. `STREAM_DISCONNECTED_DELAY` expires before `ReadLineAsync` raises exception, which means due to `WhenAny` it's not actually caught before garbage collection, since it's local to that function. 

I will add `StreamStopped` to see what it says next time it happens.

**Comment by nikop** [comment]
09.57.56.8596: Error  | --- Date : 20.7.2016 9.57.34
URL : https://stream.twitter.com/1.1/statuses/filter.json?&follow=[user ids]&stall_warnings=true
Code : 408
Error documentation description : Twitter was not able to perform your query within the Timeout limit of 10000 ms.
Error message : https://stream.twitter.com/1.1/statuses/filter.json?&follow=[user ids]&stall_warnings=true web request timed out.

09.57.56.8676: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.IO.IOException: Unable to read data from the transport connection: A blocking operation was interrupted by a call to WSACancelBlockingCall. ---> System.Net.Sockets.SocketException: A blocking operation was interrupted by a call to WSACancelBlockingCall
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.ConnectStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.Http.HttpClientHandler.WebExceptionWrapperStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Http.DelegatingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamReader.ReadBuffer()
   at System.IO.StreamReader.ReadLine()
   at AsyncExtensions.<>c__DisplayClassf.<ReadLineAsync>b__e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.IO.IOException: Unable to read data from the transport connection: A blocking operation was interrupted by a call to WSACancelBlockingCall. ---> System.Net.Sockets.SocketException: A blocking operation was interrupted by a call to WSACancelBlockingCall
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.ConnectStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.Http.HttpClientHandler.WebExceptionWrapperStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Http.DelegatingStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamReader.ReadBuffer()
   at System.IO.StreamReader.ReadLine()
   at AsyncExtensions.<>c__DisplayClassf.<ReadLineAsync>b__e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()<---

**Comment by linvi** [comment]
Sorry for not looking into this for quite a while. Have you been able to fix this problem?

**Comment by nikop** [comment]
This is still happening. I haven't had time to fix it. However I think it can be fixed with using "ContinueWith" with option OnlyOnFaulted.

**Comment by linvi** [comment]
Looking quickly online it looks like the Socket is closed from somewhere else.
Are you running various threads to run streams?

Where would you put a `.ContinueWith`?

**Comment by nikop** [comment]
There is only one stream

var requestTask = reader.ReadLineAsync();
requestTask.ContinueWith(c => { var ignored = c.Exception; }, TaskContinuationOptions.OnlyOnFaulted);

**Comment by linvi** [comment]
I am not sure that I understand how this could help for the issue.
From my understand this would simply ignore the error but in no way this would prevent the error from happening again.

Have you tried modifying Tweetinvi to see if this could potentially help?

**Comment by linvi** [comment]
Could you also tell me which type of application you are running (ASP.NET, WPF, Winform...)?

**Comment by nikop** [comment]
It's console application that acts as backend. I'm not able to share full project. This is one really hard for me too as it's very random.

My current theory is:

`var resultingTask = TaskEx.WhenAny(requestTask, TaskEx.Delay(STREAM_DISCONNECTED_DELAY)).Result;`

This goes over 90s so resultingTask is not requestTask

```
if (resultingTask != requestTask)
            {
                var twitterQueryParameter = _twitterTimeoutExceptionFactory.GenerateParameterOverrideWrapper("twitterQuery", twitterQuery);
                var twitterTimeoutException = _twitterTimeoutExceptionFactory.Create(twitterQueryParameter);
                throw (Exception)twitterTimeoutException;
}
```

This means this exception will be thrown and requestTask no longer has any reference, which means any expection will be throw by GC.

`_currentStreamReader` (same as reader) is then disposed (https://github.com/linvi/tweetinvi/blob/master/Tweetinvi.Streams/StreamTask.cs#L162)

which causes

`var requestTask = reader.ReadLineAsync();`

to throw an exception, which no longer will be caught.

I tried compiling my own version, but it wasn't successful and I haven't had time to investigate that.

**Comment by linvi** [comment]
Ok, I didn't thought about this. I will look into it more carefully and try to provide a fix before release 1.1.

**Comment by linvi** [comment]
I added the `.ContinueWith` solution to version 1.1. I hope this will solve your problem.
Reopen the issue if it does not.

Thanks for the time you took to give all the information and details.
